### PR TITLE
zfs, lvm raid, btrfs subvolumes support & some fixups

### PR DESCRIPTION
**Pull request by Lassulus**

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -180,6 +207,31 @@ let
...
+    recursiveUpdate
+    (foldl' recursiveUpdate {} (map (mount-f (q // { pool = q.name; })) x.datasets))
+    {zpool.${q.device} = ''
+      zpool list '${q.name}' >/dev/null 2>/dev/null || zpool import '${q.name}'
+    '';}
+  );
+
+  mount.zfs_filesystem = q: x: {
+      zfs.${x.mountpoint} = ''
+        if ! grep -qs '${q.pool}/${x.name} /mnt${x.mountpoint}' /proc/mounts; then
```
Comment (Mic92) [inline default.nix:223]: findmnt does not work here?

Comment (Lassulus) [reply default.nix:223]: it did not in my tests

Comment (Lassulus) [reply default.nix]: ok, now it works

File: tests/zfs.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    { config, pkgs, modulesPath, ... }:
+
+    {
+      imports = [
+        (modulesPath + "/profiles/installation-device.nix")
+        (modulesPath + "/profiles/base.nix")
+      ];
+
+      # speed-up eval
+      documentation.enable = false;
```
Comment (Mic92) [inline tests/zfs.nix:26]: This can go to a separate module that is included from every test.

File: tests/luks-lvm.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}:
+let
+  makeTest' = args:
+    makeTest args {
+      inherit pkgs;
+      inherit (pkgs) system;
+    };
+  disko-config = import ../example/luks-lvm.nix;
+  tsp-create = pkgs.writeScript "create" ((pkgs.callPackage ../. {}).create disko-config);
+  tsp-mount = pkgs.writeScript "mount" ((pkgs.callPackage ../. {}).mount disko-config);
```
Comment (Mic92) [inline tests/luks-lvm.nix:12]: Could this be a general test function that we re-ruse?
```
makeDiskoTest = diskoconfig: func: let
     tsp-create = pkgs.writeScript "create" ((pkgs.callPackage ../. {}).create disko-config);
     tsp-mount = pkgs.writeScript "mount" ((pkgs.callPackage ../. {}).mount disko-config);
     args = func { inherit tsp-create, tsp-mount; };
in
     makeTest args {
      inherit pkgs;
      inherit (pkgs) system;
    };
```

File: default.nix [hunk 4]
```diff
@@ -125,6 +125,31 @@ let
...
+      -V ${x.size} \
+      ${lib.optionalString (isAttrs x.options or null) (concatStringsSep " " (mapAttrsToList (n: v: "-o ${n}=${v}") x.options))}
+    udevadm trigger --subsystem-match=block; udevadm settle
+    ${create-f { device = "/dev/zvol/${q.pool}/${x.name}"; } x.content}
+  '';
+
+  create.zpool = q: x: ''
+    zpool create ${q.name} \
+      ${lib.optionalString (!isNull x.mode || x.mode != "stripe") x.mode} \
+      ${lib.optionalString (isAttrs x.options or null) (concatStringsSep " " (mapAttrsToList (n: v: "-o ${n}=${v}") x.options))} \
```
Comment (0x4A6F) [inline default.nix:148]: Is there a way to set filesystem options with -O on the root dataset?

File: tests/luks-lvm.nix (hunk 3 above)
Comment (Lassulus) [reply tests/luks-lvm.nix:12]: we probably need to define the number of disks too, but I guess we can generalize that

File: default.nix (hunk 4 above)
Comment (Lassulus) [reply default.nix:163]: do we need that? what would be the difference to just -o on the zpool?

File: tests/luks-lvm.nix (hunk 3 above)
Comment (Lassulus) [reply tests/luks-lvm.nix:12]: actually we could get the number of disks from the diskoConfig

File: default.nix (hunk 4 above)
Comment (0x4A6F) [reply default.nix:163]: `zpool create` will create the root zfs dataset, without -O these dataset properties can only be set with `zfs set` afterwards. `zpool create -o` does only set pool properties and not dataset properties.

File: default.nix [hunk 5]
```diff
@@ -52,10 +52,10 @@ let
     boot.initrd.luks.devices.${x.name}.device = q.device;
   } // config-f { device = "/dev/mapper/${x.name}"; } x.content;
 
-  config.lv = q: x:
+  config.lvm_lv = q: x:
     config-f { device = "/dev/mapper/${q.vgname}-${q.name}"; } x.content;
```
Comment (ajs124) [inline default.nix:56]: escaped ihr das richtig, wenn der name `-` enthält?

File: default.nix [hunk 6]
```diff
@@ -98,15 +98,23 @@ let
...
+    pvcreate ${q.device}
+    LVMDEVICES_${x.vg}="''${LVMDEVICES_${x.vg}:-}${q.device} "
+  '';
+
+  create.lvm_lv = q: x: ''
+    lvcreate \
+      ${if hasInfix "%" x.size then "-l" else "-L"} ${x.size} \
+      -n ${q.name} \
+      ${lib.optionalString (!isNull x.lvm_type or null) "--type=${x.lvm_type}"} \
+      ${q.vgname}
```
Comment (ajs124) [inline default.nix:111]: gibt es da keine extraargs oder sowas? was wenn ich `--stripes` oder sowas angeben will?

Comment (ajs124) [reply default.nix:111]: einen thinpool und thin lvs darauf kann ich hiermit relativ sicher auch nicht erstellen :thinking:

File: default.nix [hunk 7]
```diff
@@ -38,42 +39,65 @@ let
     };
   };
 
+  config.zfs_filesystem = q: x: {
```
Comment (mweinelt) [inline default.nix:42]: q? x?

Comment (Lassulus) [reply default.nix:42]: q is the state upwards and x the configuration downwards

**Pull request merged**